packages/opentelemetry-metrics/src/BaseObserverMetric.ts

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as api from '@opentelemetry/api';
+import { Resource } from '@opentelemetry/resources';
+import { BoundObserver } from './BoundInstrument';
+import { Batcher } from './export/Batcher';
+import { MetricKind, MetricRecord } from './export/types';
+import { Metric } from './Metric';
+import { ObserverResult } from './ObserverResult';
+
+const NOOP_CALLBACK = () => {};
+
+/**
+ * This is a SDK implementation of Base Observer Metric.
+ * All observers should extend this class
+ */
+export abstract class BaseObserverMetric extends Metric<BoundObserver>
+  implements api.BaseObserver {
+  protected _callback: (observerResult: api.ObserverResult) => void;
+
+  constructor(
+    name: string,
+    options: api.MetricOptions,
+    private readonly _batcher: Batcher,
+    resource: Resource,
+    metricKind: MetricKind,
+    callback?: (observerResult: api.ObserverResult) => void
+  ) {
+    super(name, options, metricKind, resource);
+    this._callback = callback || NOOP_CALLBACK;
```

Comment:
there will always be a way to blow up the module by using it incorrectly. sending the wrong data type is a user error and an error is the correct behavior IMO

Reply:
That's fair. Maybe instead of defaulting the value we could eagerly throw an error like `new Error('Expected a function for callback parameter, instead got ' + typeof callback)` if the type doesn't match? Waiting until the first call to the mistaken data doesn't seem like the best ergonomically.

Reply:
we generally doesn't check so strictly and trying to prevent all possible input errors for none ts. If we are going to change it I would say we should probably revisit our general approach for user inputs validation. Although I would be carefull with defencing too much as it will add extra overhead in code. Not sure if we want that @dyladan ?

Reply:
This is hard to balance as we generally try not to throw an error in any case unless it is required (this is a requirement from spec). If a user inputs bad input we obviously will probably throw when that input is called like a function but isn't one, but that can't always be avoided. If we were going to try to be defensive here, I would say logging and making this a NOOP metric would probably be the most spec-compliant way. Since this is not a hot code path, I think spending a couple cycles to check the input is probably not a problem, but I would try not to throw if possible.

Reply:
That sounds like a good compromise!

Also, sorry, I haven't seen any previous discussion about these kinds of conventions, is there a place to catch up on them?

Reply:
This particular one comes from the spec. I have been thinking about making a dev guide though since we have an infusion of new people recently with the various intern programs starting.